src/solver/optimize.cpp

```diff
@@ -1,14 +1,294 @@
+#include <math.h>
+#include <string>  
+#include <iostream> 
+#include <sstream> 
 #include "../components/token.hpp"
 #include "../components/order.hpp"
 #include "../components/amm.hpp"
 #include "./optimize.hpp"
+#include "../../external_libs/json.hpp"
 
-void solve_auction(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms)
+
+// auxiliary function (used temporarily) for computing x^n using the boost library data types
+boost::multiprecision::mpf_float custom_pow(boost::multiprecision::mpz_int x, int n)
+{
+    int i;
+    boost::multiprecision::mpf_float res = 1;
+    int m = (n > 0) ? n : (-n);
+
+    for (i = 0; i < m; i++)
+        res = res * x;
+    if (n > 0)
+        return res;
+    else
+        return (1/res);
+}
+
+// main function that processes a sell order, checks if it can be matched against a single liquidity pool
+// and if so, it also checks whether it gives a strictly better objective
+void process_sell_order(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, int i, ExecutedOrder &best_order)
 {
-    // TO BE COMPLETED
-    int *vertices;
-    int **edges;
+   int j, num_amms;
+    boost::multiprecision::mpz_int x0, y0, x1, y1, x_, y_, x;
+    boost::multiprecision::mpf_float rational_x1;
+    boost::multiprecision::mpf_float rational_x;
+    boost::multiprecision::mpf_float temp;
+
     
+    x_ = orders[i].buy_amount;
+    y_ = orders[i].sell_amount;
+    num_amms = amms.size();
+
+    for (j = 0; j < num_amms; j++) {
+        x0 = 0;
+        y0 = 0;
+        if (amms[j].sell_token_index == orders[i].sell_token_index)
+            x0 = amms[j].sell_reserve_amount;
+        if (amms[j].buy_token_index == orders[i].sell_token_index)
+            x0 = amms[j].buy_reserve_amount;
+        if (amms[j].sell_token_index == orders[i].buy_token_index)
+            y0 = amms[j].sell_reserve_amount;
+        if (amms[j].buy_token_index == orders[i].buy_token_index)
+            y0 = amms[j].buy_reserve_amount;
+
+        if ((x0 <= 0) || (y0 <= 0))
+            continue;
+
+        // at this point, we have found a "matching" AMM
+        // we must satisfy x1 * y1 >= x0 * y0, where (x1,y1) is the new pool balance and (x0, y0) is the original pool balance
+        // Since it is a fill-or-kill sell order, we have x1 = x0 + 0.997y_
+        // and y1 = y0 - x
+        // Thus, maximally exploiting the pool, we get that y1 = x0 * y0 / x1,
+        // which implies that x = y0 * (1 - x0/x1)
+        rational_x1 = static_cast<boost::multiprecision::mpf_float>(x0) + 0.997 * static_cast<boost::multiprecision::mpf_float>(y_);
```

Comment:
Where is the 0.997 coming from?

Reply:
Oh sorry, my bad. I didn't realize I had hardwired this number in the code. I replaced it with (1-fee) which should be the right thing to do for a Uniswap Constant Product AMM.